core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -147,7 +149,7 @@ public String getRedirectUrl(@Nonnull Token token, @Nonnull String restOfPath) {
             // Unsure whether this can happen -- just be safe here
             restOfPath = "/" + restOfPath;
         }
-        return resourceRootUrl + getUrlName() + "/" + token.encode() + restOfPath;
+        return resourceRootUrl + getUrlName() + "/" + token.encode() + Arrays.stream(restOfPath.split("[/]")).map(Util::rawEncode).collect(Collectors.joining("/"));
```

Comment:
This is going to drop a trailing `/` so that is no good. It will also convert `"/"` to `""`.

Reply:
Added a test that hopefully demonstrates that this does not matter.